examples/hyper_server.rs

```diff
@@ -43,50 +50,33 @@ async fn main() -> Result<(), BoxedError> {
 
 /// Handle incoming HTTP Requests.
 async fn handler(req: Request<Body>) -> Result<hyper::Response<Body>, BoxedError> {
-	// If the request is asking to be upgraded to a websocket connection, we do that,
-	// and we handle the websocket connection (in this case, by echoing messages back):
-	if is_upgrade_request(&req) {
-		let (res, on_upgrade) = upgrade_to_websocket(req)?;
-		tokio::spawn(async move {
-			if let Err(e) = websocket_echo_messages(on_upgrade).await {
-				eprintln!("Error upgrading to websocket connection: {}", e);
-			}
-		});
-		Ok(res)
-	}
-	// Or, we can handle the request as a standard HTTP request:
-	else {
-		Ok(Response::new(Body::from("Hello HTTP!")))
-	}
-}
-
-/// Return the response to the upgrade request, and a way to get hold of the underlying TCP stream
-fn upgrade_to_websocket(req: Request<Body>) -> Result<(Response<Body>, hyper::upgrade::OnUpgrade), handshake::Error> {
-	let key = req.headers().get("Sec-WebSocket-Key").ok_or(handshake::Error::InvalidSecWebSocketAccept)?;
-	if req.headers().get("Sec-WebSocket-Version").map(|v| v.as_bytes()) != Some(b"13") {
-		return Err(handshake::Error::HeaderNotFound("Sec-WebSocket-Version".into()));
+	match upgrade_request(&req) {
+		// The upgrade was successful, so we return the response and start up a server to take charge of the socket:
+		Ok(Upgraded { response, server_configuration }) => {
+			tokio::spawn(async move {
+				let on_upgrade = hyper::upgrade::on(req);
```

Comment:
I'll add a bunch more comments to explain what's going on here!

Basically, `on_upgrade.await?` gives back as AsyncRead + AsyncWrite object that reads/writes to the underlying stream, and we can hand this to Soketto to run the websocket protocol on that stream.

I wonder whether `upgrade_request` is the right name; maybe `negotiate_upgrade` or similar would be better, because that's really what it's doing; agreeing with the client on the parameters of the upgrade and then expecting the stream to be used for that next!

Reply:
+1 for `negotiate_upgrade`!